README
NAME
    Term::FormatColumns - Format lists of data into columns across the
    terminal's width

VERSION
    version 0.006

SYNOPSIS
        use Term::FormatColumns qw( format_columns );
        my @list = 0..1000;
        print format_columns @list;

DESCRIPTION
    This module will take a list and format it into columns that stretch
    across the current terminal's width, much like the output of ls(1).

    If the filehandle is not attached to a tty, will simply write one column
    of output (again, like ls(1)).

SUBROUTINES
  format_columns
        my $string = format_columns @array;

    Format the list of data for STDOUT. Returns a single string formatted
    and ready for output.

  format_columns_for_fh
        my $string = format_columns_for_fh $fh, @array;
        my $string = format_columns_for_fh STDOUT, @array;

    Format the given data for the given filehandle. If the filehandle is
    attached to a tty, will get the tty's width to determine how to format
    the data.

  format_columns_for_width
        my $string = format_columns_for_width 78, @array;

    Format the given data for the given width. This allows you to use this
    module without being attached to a known/knowable terminal.

AUTHOR
    Doug Bell <preaction@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Doug Bell.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Changelog for Term-FormatColumns

0.006 2015-01-02T23:29:17
 - remove the _ prototype to fix 5.8
 - remove s///r to allow Perl prior to 5.14
 - update module for Pod::Weaver docs
 - add authordep on Pod::Weaver::Section::Contributors
 - use updated Git::Contributors dzil plugin
 - add contributors dzil plugin for proper attribution
 - rebuild build artifacts to trigger a travis build
 - add travis-ci for testing
 - update dist.ini to current best practices
 - add missing prereqs
 - Fix whitespace errors, use qw and __END__
 - Remove unnecessary prototypes
 - Use strict and warnings
 - Add newline at end of format_columns_for_fh non-tty output
 - Ignore ANSI SGR sequences when computing string length
 - Pad lists shorter than rows * cols with empty strings
 - update weaver.ini from new skeleton
 - SurgicalPodWeaver -> PodWeaver

0.005 2014-04-08T03:08:58
 - fix name resolution issue with sub
 - add build artifacts from dist.ini, for easier contributions
 - update to my standard dist.ini

0.004 2013-01-11T02:48:12
 - update README from POD
 - clarify the docs a little bit
 - use glob in the prototype instead of scalars, recommended by Joel Berger

0.003 2013-01-11T01:20:38
 - update README from POD
 - add an actual test routine!
 - don't pad out the last column with unnecessary whitespace
 - add a simple compile test
 - always end with a \n
 - add format_columns_for_width in case we don't have a tty
 - update e-mail address

0.002 2012-12-30T22:30:16
 - updated readme from POD
 - rename the subs to be a more consistent with the module name
 - Sub::Exporter wants a list, not an arrayref

0.001 2012-12-28T04:08:00
 - Initial commit



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.023.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "Format lists of data into columns across the terminal's width",
  "dist_author" => [
    "Doug Bell <preaction\@cpan.org>"
  ],
  "dist_name" => "Term-FormatColumns",
  "dist_version" => "0.006",
  "license" => "perl",
  "module_name" => "Term::FormatColumns",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "List::MoreUtils" => "0.33",
    "List::Util" => 0,
    "POSIX" => 0,
    "Sub::Exporter" => "0.984",
    "Symbol" => 0,
    "Term::ReadKey" => "2.30",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Test::Compile" => 0,
    "Test::More" => "0.88"
  }
);


my %fallback_build_requires = (
  "Module::Build" => "0.28",
  "Test::Compile" => 0,
  "Test::More" => "0.88"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Term/FormatColumns.pm
package Term::FormatColumns;
# ABSTRACT: Format lists of data into columns across the terminal's width

use strict;
use warnings;

use Sub::Exporter -setup => [
    exports => (
        qw/format_columns format_columns_for_fh format_columns_for_width/,
    ),
];

use Term::ReadKey qw( GetTerminalSize );
use List::Util qw( max );
use List::MoreUtils qw( part each_arrayref );
use POSIX qw( ceil );
use Symbol qw(qualify_to_ref);

# Find the length of a string as displayed on the terminal, ignoring any ANSI
# escape sequences.
sub _term_length {
    my ( $str ) = @_;
    $str =~ s/\x1b\[[0-9;]+m//g;
    return length $str;
}

=sub format_columns

    my $string = format_columns @array;

Format the list of data for STDOUT. Returns a single string formatted and ready for output.

=cut

sub format_columns {
    return format_columns_for_fh( \*STDOUT, @_ );
}

=sub format_columns_for_fh

    my $string = format_columns_for_fh $fh, @array;
    my $string = format_columns_for_fh STDOUT, @array;

Format the given data for the given filehandle. If the filehandle is attached to a tty,
will get the tty's width to determine how to format the data.

=cut

sub format_columns_for_fh(*@) {
    my $fh = qualify_to_ref( shift, caller );
    my @data = @_;

    # If we're not attached to a terminal, one column, seperated by newlines
    if ( !-t $fh ) {
        return join "\n", @data, '';
    }

    # We're attached to a terminal, print column-wise alphabetically to fit the
    # terminal width
    my ( $term_width, undef, undef, undef ) = GetTerminalSize();
    return format_columns_for_width( $term_width, @data );
}

=sub format_columns_for_width

    my $string = format_columns_for_width 78, @array;

Format the given data for the given width. This allows you to use this module without
being attached to a known/knowable terminal.

=cut

sub format_columns_for_width {
    my ( $term_width, @data ) = @_;
    my $max_width = max map { _term_length( $_ ) } @data;
    $max_width += 2; # make sure at least two spaces between data values
    my $columns = int( $term_width / $max_width );
    if ( $columns <= 1 ) {
        # Only one column, let the terminal handle things
        return join "\n", @data, ''; # Add a \n to the end
    }
    my $output = '';
    my $column_width = int( $term_width / $columns );
    my $rows = ceil( @data / $columns );
    push @data, ('') x ($rows * $columns - @data); # Pad data with empty strings
    my @index = part { int( $_ / $rows ) } 0..$#data;
    my $iter = each_arrayref @index;
    while ( my @row_vals = $iter->() ) {
        my @cells = map { $data[$_] } @row_vals;
        my $last_cell = pop @cells;
        for (@cells) {
            my $length = _term_length( $_ );
            $output .= $_;
            $output .= ' ' x ($column_width - $length);
        }
        $output .= $last_cell . "\n";
    }
    return $output;
}

1;
__END__

=head1 SYNOPSIS

    use Term::FormatColumns qw( format_columns );
    my @list = 0..1000;
    print format_columns @list;

=head1 DESCRIPTION

This module will take a list and format it into columns that stretch across the
current terminal's width, much like the output of ls(1).

If the filehandle is not attached to a tty, will simply write one column of output
(again, like ls(1)).


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.023.
use strict;
use warnings;



use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Format lists of data into columns across the terminal's width",
  "AUTHOR" => "Doug Bell <preaction\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.28"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "DISTNAME" => "Term-FormatColumns",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Term::FormatColumns",
  "PREREQ_PM" => {
    "List::MoreUtils" => "0.33",
    "List::Util" => 0,
    "POSIX" => 0,
    "Sub::Exporter" => "0.984",
    "Symbol" => 0,
    "Term::ReadKey" => "2.30",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Test::Compile" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.006",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "List::MoreUtils" => "0.33",
  "List::Util" => 0,
  "Module::Build" => "0.28",
  "POSIX" => 0,
  "Sub::Exporter" => "0.984",
  "Symbol" => 0,
  "Term::ReadKey" => "2.30",
  "Test::Compile" => 0,
  "Test::More" => "0.88",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



